clang/lib/Headers/amdgpuintrin.h

```diff
@@ -0,0 +1,155 @@
+//===-- amdgpuintrin.h - AMDPGU intrinsic functions -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __AMDGPUINTRIN_H
+#define __AMDGPUINTRIN_H
+
+#ifndef __AMDGPU__
+#error "This file is intended for AMDGPU targets or offloading to AMDGPU"
+#endif
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#if !defined(_DEFAULT_FN_ATTRS)
+#if defined(__HIP__) || defined(__CUDA__)
+#define _DEFAULT_FN_ATTRS __attribute__((device))
+#else
+#define _DEFAULT_FN_ATTRS
+#endif
+#endif
+
+#pragma omp begin declare target device_type(nohost)
+#pragma omp begin declare variant match(device = {arch(amdgcn)})
```

Comment:
How much do we have to worry about users defining any of these as macros prior to including the system header?

Reply:
These should be ignored outside of OpenMP mode, but if someone does like `#define nohost "foo"` it will definitely break.

Reply:
Is there a way we can guard against that? Does OpenMP allow for things like `#pragma __omp__ __begin__ ... `?

Reply:
I could probably use `_Pragma("..")` instead.

Reply:
Done